runtime/runtime-params-estimator/continuous-estimation/run_estimation.sh

```diff
@@ -0,0 +1,77 @@
+#!/bin/bash
+set -e
+
+SOURCE_REPO_HOME=`git rev-parse --show-toplevel`
+
+if [[ -z "${SQLI_DB}" ]]; then
+  echo '$SQLI_DB is not defined, please set it as environment variable before running this script.'
+  exit
+fi
+
+if [[ -z "${ESTIMATOR_NEAR_HOME}" ]]; then
+  echo '$ESTIMATOR_NEAR_HOME is not defined, will use /tmp/estimator_home as near home'
+  ESTIMATOR_NEAR_HOME="/tmp/estimator_home"
+  mkdir -p $ESTIMATOR_NEAR_HOME
+fi
+
+if [[ -z "${REPO_UNDER_TEST}" ]]; then
+  echo '$REPO_UNDER_TEST is not defined, will use repository that the script is running from.'
+else
+  echo "Estimating repository at ${REPO_UNDER_TEST}"
+  cd $REPO_UNDER_TEST
+fi
+
+# Build warehouse binary
+pushd "${SOURCE_REPO_HOME}"
+cargo build --release
+# Some users have CARGO_TARGET_DIR pointing to a custom target directory
+if [[ ${CARGO_TARGET_DIR} ]]; then
+  WAREHOUSE_BINARY="${CARGO_TARGET_DIR}/release/estimator-warehouse"
+else
+  WAREHOUSE_BINARY="`git rev-parse --show-toplevel`/target/release/estimator-warehouse"
+fi
+popd
+
+# Ensure full optimization
+export CARGO_PROFILE_RELEASE_LTO=fat
+export CARGO_PROFILE_RELEASE_CODEGEN_UNITS=1
+
+# Rebuild test contract
+pushd "${SOURCE_REPO_HOME}/runtime/runtime-params-estimator/test-contract"
+bash build.sh
+popd
+
+# Build estimator
+# Again, some users have CARGO_TARGET_DIR pointing to a custom target directory
+if [[ ${CARGO_TARGET_DIR} ]]; then
+  ESTIMATOR_BINARY="${CARGO_TARGET_DIR}/release/runtime-params-estimator"
+else
+  ESTIMATOR_BINARY="`git rev-parse --show-toplevel`/target/release/runtime-params-estimator"
+fi
+cargo build --release -p runtime-params-estimator --features required;
+
+
+#  Actual estimations
+GIT_HASH=`git rev-parse HEAD`
+COMMON_ARGS="--iters 5 --warmup-iters 1 --json-output"
+# This has problems in qemu, leading to OOM
+USE_LARGE_STATE="--accounts-num 1000000 --home ${ESTIMATOR_NEAR_HOME}"
+# Options that are only relevant for time based estimations
+TIME_OPTIONS="--drop-os-cache"
+# Options that are only relevant for icount based estimations
+ICOUNT_OPTIONS="--docker --full "
+
+# Run time measurements outside docker and with sudo if possible.
+if [ "$EUID" -ne 0 ]; then
+  echo "Running as non-root, storage related costs might be inaccurate because OS caches cannot be dropped"
+  TIME_OPTIONS=""
+fi
+
+IMPORT_OPTIONS="--db ${SQLI_DB} import --commit-hash=${GIT_HASH} "
+
+"${ESTIMATOR_BINARY}" ${COMMON_ARGS} ${USE_LARGE_STATE} ${TIME_OPTIONS} --metric time \
+    | "${WAREHOUSE_BINARY}" ${IMPORT_OPTIONS};
+
+# Run icount measurements inside docker container.
+"${ESTIMATOR_BINARY}" ${COMMON_ARGS} ${ICOUNT_OPTIONS} --metric icount \
+    | "${WAREHOUSE_BINARY}" ${IMPORT_OPTIONS};
```

Comment:
Can we pull this bash script into the wherehouse binary itself? Not a huge fan of mixing bash and Rust. You can use https://github.com/matklad/xshell to make the rust version of this relatively concise. 

Though, no strong opinion here -- if you prefer bash, then so be it!

Reply:
Xshell looks neat, I will give it a try and see how it goes. If we can remove bash from the equation, that seems overall positive.

Generally speaking though, for me personally bash and Rust are fundamentally different tools that often complement each other perfectly. (Rust) binaries have these clearly defined interfaces with CLI arguments and stdin/stdout. Bash is designed to use this interface effectively in clear and concise code. Rust's strong type system does not help much with that task and the re-compiling for every constant string value change is actively harming productivity.

Reply:
> Bash is designed to use this interface effectively in clear and concise code. Rust's strong type system does not help much with that task and the re-compiling for every constant string value change is actively harming productivity.

Yeah, that's true. My argument here is that, *despite* bash being a better tool for this particular job, using Rust might be overall better, because we'll stick to a single language/ecosystem. Basically, this: https://twitter.com/id_aa_carmack/status/989951283900514304

No strong opinion though!

Reply:
Thanks to xshell, I was able to remove bash from the process without too much of a hassle. Although it was more work than the bash script and I am not sure which is more readable, I do really like the fact that now I can do `estimator-warehouse estimate` instead of calling a bash script. Overall it was worth it IMO.

Before merging this, can you please take a look at the new file `estimate.rs` that replaced `run_estimation.sh`? It feels like I probably missed better ways to do it in a few places, so if you see something obvious it would be appreciated.